core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -414,7 +414,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * should generally not be used for releasing resources. For higher-level
    * logic built on `ensuring`, see [[ZIO#bracket]].
    */
-  final def ensuring(finalizer: UIO[_]): ZIO[R, E, A] =
+  final def ensuring[R0 <: R](finalizer: ZIO[R0, Nothing, _]): ZIO[R0, E, A] =
```

Comment:
I don't think this is sound because finalizers are run non-deterministically. They could be run "in order", after the main body, but the main body could have thrown.

This is why the code has `ensuringR`, which comes with a caveat: it uses the environment at the time the `ensuringR` is invoked, which is _not_ (necessarily) the same as the environment at the time the body is invoked.

Reply:
We need to think really carefully about this. _If_ it is sound, it could be a good idea. I'm just not sure it is. Indeed, I believe that `provide` uses `ensuring` to ensure the right environment is restored on the way out.

Reply:
Ha, I missed `ensuringR` and the caveat in the scaladoc 👍 Indeed `provide` uses `ensuring`. It feels like a rabbit hole, I'm happy to play with it in a separate PR 😄